### ISSUE TITLE
Project API Fails to Pull Connector Modules Intermittently

### ISSUE THREAD

**Issue by miyurud** [ballerina-platform/ballerina-lang#30642]
**Description:**
We have observed a rare intermittent error thrown by a Java Micro Service which parses Ballerina Build projects. Basically the connector does not get pull from the Ballerina Central. This happens when a connector gets pulled for the very first time (may be first few requests). Once the connector got pulled the error does not get thrown again because the connector module get cached. The logs we got in the Java Micro Service is as follows,

```
2021-05-15 10:39:24 DEBUG ActionInvocationParser:99 - time elapsed 3299ms 
2021-05-15 10:39:24 WARN  ProgramAnalyzerNodeVisitor:411 - Returning since we are only interested in new constructor expressions. Found expression kind : CHECK_ACTION 
 ballerinax/googleapis.sheets:0.99.8 pulled from central successfully 
 ballerinax/googleapis.sheets:0.99.8 [central.ballerina.io -> home repo]   47% [ ballerinax/googleapis.sheets:0.99.8 [central.ballerina.io -> home repo]  100% [ 
2021-05-15 10:32:02 ERROR ProgramAnalyzerController:140 - Analyzer fails to detect any endpoints or action invocations 
2021-05-15 10:32:02 DEBUG ActionInvocationParser:99 - time elapsed 1099ms 
2021-05-15 10:32:02 WARN  ProgramAnalyzerNodeVisitor:47 - Unable to find endpoint initiation for remote call: appendCell at position:12:29. null 
2021-05-15 10:32:02 WARN  ProgramAnalyzerNodeVisitor:411 - Returning since we are only interested in new constructor expressions. Found expression kind : CHECK_ACTION 
2021-05-15 10:32:02 DEBUG ActionInvocationParser:72 - -->invalid remote method call: expected a client object, but found 'other' 
2021-05-15 10:32:02 DEBUG ActionInvocationParser:72 - -->cannot infer type of the object from '(other|error)' 
2021-05-15 10:32:02 DEBUG ActionInvocationParser:72 - -->unknown type 'Client' 
2021-05-15 10:32:02 DEBUG ActionInvocationParser:72 - -->undefined module 'sheets' 
2021-05-15 10:32:02 DEBUG ActionInvocationParser:72 - -->cannot resolve module 'ballerinax/googleapis.sheets as sheets' 
2021-05-15 10:30:43 DEBUG ActionInvocationParser:99 - time elapsed 1538ms 
2021-05-15 10:30:43 WARN  ProgramAnalyzerNodeVisitor:47 - Unable to find endpoint initiation for remote call: respond at position:32:29. null 
2021-05-15 10:30:43 WARN  ProgramAnalyzerNodeVisitor:411 - Returning since we are only interested in new constructor expressions. Found expression kind : CHECK_ACTION 
2021-05-15 10:30:43 WARN  ProgramAnalyzerNodeVisitor:411 - Returning since we are only interested in new constructor expressions. Found expression kind : CHECK_ACTION 
2021-05-15 10:29:58 DEBUG ActionInvocationParser:99 - time elapsed 1496ms 
2021-05-15 10:29:58 WARN  ProgramAnalyzerNodeVisitor:411 - Returning since we are only interested in new constructor expressions. Found expression kind : CHECK_ACTION 
2021-05-15 10:28:32 DEBUG ActionInvocationParser:99 - time elapsed 1091ms 
```

In the above log the log lines which contain `-->` are diagnostic logs printed by Ballerina Project API. The attempt around `2021-05-15 10:32:02` was failed. However, the second attempt around `2021-05-15 10:39:24` was successful and the connector got pulled in to local Ballerina cache. Note that both these attempts were externally triggered. When the first error got thrown there were no re-try attempts were made to pull the connector.

**Steps to reproduce:**
This could be reproduced if we create a Java application which uses Ballerina Project API to build a Ballerina build project. Every time before we run the Java project we need to clean the `.ballerina` cache. Running the above Java app multiple times could reproduce the above mentioned issue.

**Affected Versions:**
Ballerina SL Alpha 5.3

**OS, DB, other environment details and versions:**
Ubuntu 20.04.2.0, JDK 11

**Comment by miyurud** [comment]
@Ibaqu I have attached a sample Java project `partial-compiler` which can be used to reproduce the above issue. Please follow the steps mentioned below,

- Extract `partial-compiler.zip` and cd to `partial-compiler`
- invoke `mvn clean install`
- run `java -cp target/program-analyzer-0.4.2-SNAPSHOT-jar-with-dependencies.jar controller.ballerina.parser.ActionInvocationParser` from the `partial-compiler` directory

Try running the last step above in a for loop, since this is intermittent issue you might get it reproduced if you run it in multiple times.

[partial-compiler.zip](https://github.com/ballerina-platform/ballerina-lang/files/6771672/partial-compiler.zip)

**Comment by Ibaqu** [comment]
I have attempted to reproduce the issue with the provided resources, specifically the partial-compiler project using version `jballerina-tools-2.0.0-alpha8-20210616-215600-e71bbd28` as provided in the project. 

I also a separate gradle project with the same classes and dependencies, that creates the `.ballerina` cache in the target directory. A script runs the project and deletes the cache in an attempt to run it multiple times in hopes of reproducing the issue. 

Unfortunately, I can't seem to reproduce the invalid method call as specified in the issue. I have also monitored the PackageCache instance for the `packageMap` and `pacakgeSymbolMap` to see if there were any packages that were not getting resolved but to no avail. 

Since we cannot rule out certain conditions I will keep this issue open till we find a way of successfully reproducing the problem
